File: src/graph/togological-sort/topological-sort.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

void add_edge(vector<vector<int>>& arr, vector<int>& incoming, const int src, const int dest)
{
    arr[src].push_back(dest);
    incoming[dest]++;
}

bool topological_sort(vector<vector<int>>& arr, vector<int>& incoming)
{
    queue<int> task_q;
    vector<int> result;
    int size = (int)arr.size();

    for (int i = 0; i < size; i++) // Enqueue vertex if it does not have any incoming edges.
        if (!incoming[i]) task_q.push(i);

    while (!task_q.empty()) // Do until the queue is empty.
    {
        int v = task_q.front();
        task_q.pop();
        result.push_back(v);

        int dest_size = (int)arr[v].size();
        for (int i = 0; i < dest_size; i++)
        {
            if (--incoming[arr[v][i]] == 0) // Enqueue vertex under the same condition.
                task_q.push(arr[v][i]);
        }
    }
    if (result.size() != size) // Returns false if the graph has a cycle.
        return false;
    for (int i = 0; i < size; i++)
        cout << result[i] << ' ';
    cout << '\n';

    return true;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0); cout.tie(0);
    int n, src, dest;

    cin >> n; // Input the number of vertices.
    vector<vector<int>> arr(n);
    vector<int> incoming(n);

    while (1)
    {
        cin >> src;
        if (src == -1) break;
        cin >> dest;

        add_edge(arr, incoming, src, dest);
    }
    topological_sort(arr, incoming);

    return 0;
}

/*
Input:
5
0 1
0 2
0 3
1 2
2 3
4 3
4 2
-1

Output:
0 4 1 2 3
*/

File: src/graph/topological-sort/topologicalSort.cs
using System;
using System.Collections.Generic;

namespace TSort
{
    public class CycleGraphException : Exception
    {
        public CycleGraphException()
        {
            Console.Write("Exception - CycleGraphException : ");
        }
    }

    public static class MainApp
    {
        private static List<List<int>> _digraph;
        private static List<int> _incoming;
        private static int _n;

        public static bool AddEdge(int src, int dest)
        {
            try
            {
                _digraph[src].Add(dest);
                _incoming[dest]++;
            }
            catch (ArgumentOutOfRangeException)
            {
                Console.WriteLine("One or more vertex number are out of range.");
                return false;
            }

            return true;
        }

        public static void TopologicalSort(out List<int> rt)
        {
            rt = new List<int>();
            var vert = new Queue<int>();

            for (int i = 0; i < _n; i++) // Enqueue nodes that does not have any incoming edges.
                if (_incoming[i] == 0)
                    vert.Enqueue(i);

            while (vert.Count > 0)
            {
                int v = vert.Dequeue();
                rt.Add(v);
                
                for (int i = 0; i < _digraph[v].Count; i++)
                    if (--_incoming[_digraph[v][i]] == 0) // Remove incoming edge and Add nodes which have no incoming edges.
                        vert.Enqueue(_digraph[v][i]);
            }
            if (rt.Count != _digraph.Count) throw new CycleGraphException(); // Throws an exception that the digraph has at least one cycle.
        }

        public static void Main()
        {
            _n = int.Parse(Console.ReadLine());
            _digraph = new List<List<int>>(new List<int>[_n]);
            for (int i = 0; i < _n; i++)
                _digraph[i] = new List<int>();
            _incoming = new List<int>(new int[_n]);

            while (true)
            {
                string[] input = Console.ReadLine().Split(' ');
                int src = int.Parse(input[0]);
                if (src == -1) break;

                int dest = int.Parse(input[1]);
                AddEdge(src, dest);
            }
            
            try
            {
                TopologicalSort(out List<int> result);
                Console.Write("Sorted vertices: ");
                foreach (var elements in result)
                    Console.Write($"{elements} ");
                Console.WriteLine();
            }
            catch (CycleGraphException)
            {
                Console.WriteLine("This digraph has at least one cycle.");
            }
        }
    }
}